### PR TITLE
Improved Test Fixtures

### PR DESCRIPTION
**Pull request by jonasBoss**
This allows us to specify the exact device to which a InputEvent is sent.
Previously it was only possible to specify the Group, which was limiting for some tests.

### CODE REVIEW

File: tests/test.py [hunk 1]
```diff
@@ -262,79 +293,190 @@ def read_write_history_pipe():
...
+            self.dev_input_event15,
+            self.dev_input_event20,
+            self.dev_input_event30,
+            self.dev_input_event31,
+            self.dev_input_event40,
+            self.dev_input_event51,
+        ]
+        self._dynamic_fixtures = {}
+
+    def __getitem__(self, item: str) -> Fixture:
```
Comment (sezanzeb) [inline tests/test.py:350]: Please add a docstring that explains `item`

File: tests/test.py [hunk 2]
```diff
@@ -170,21 +180,36 @@ def read_write_history_pipe():
...
+    name: str = "unset"
+    info: evdev.device.DeviceInfo = evdev.device.DeviceInfo(None, None, None, None)
+    phys: str = "unset"
+    group_key: Optional[str] = None
+
+    def __hash__(self):
+        return hash(self.path)
+
+
+class _Fixtures:
```
Comment (sezanzeb) [inline tests/test.py:197]: Docstring please.

It seems `_Fixtures` is some sort of container. Is it like a dict?

File: tests/test.py [hunk 3]
```diff
@@ -262,79 +293,190 @@ def read_write_history_pipe():
...
+    @property
+    def foo_device_2_keyboard(self):
+        return self["/dev/input/event10"]
+
+    @property
+    def foo_device_2_13(self):
+        return self["/dev/input/event13"]
+
+    @property
+    def foo_device_2_qux(self):
```
Comment (sezanzeb) [inline tests/test.py:403]: is `self.foo_device_2_qux` the same as `self.dev_input_event14`?

File: tests/test.py [hunk 4]
```diff
@@ -823,13 +963,13 @@ def prepare_presets():
...
 
-    time.sleep(0.05)
+    time.sleep(0.1)
     preset2 = Preset(get_preset_path("Foo Device", "preset2"))
     preset2.add(get_key_mapping(combination="1,3,1"))
     preset2.add(get_key_mapping(combination="1,4,1"))
     preset2.save()
 
-    time.sleep(0.05)  # make sure the timestamp of preset 3 is the newest
+    time.sleep(0.1)  # make sure the timestamp of preset 3 is the newest
```
Comment (sezanzeb) [inline tests/test.py:972]:
```suggestion
    # make sure the timestamp of preset 3 is the newest
    # (probably for the GUI to load this preset first)
    time.sleep(0.1)
```

**Review by sezanzeb — CHANGES_REQUESTED** [review]
Very nice. It improves the readability of tests a lot

File: tests/test.py (hunk 2 above)
Comment (jonasBoss) [reply tests/test.py:197]: It is just a class with a bunch of class variables.
It defines all the standard fixtures and provides access to them via the path (`fixtures["/dev/input/eventX"]`) the attribute name (`fixtures.dev_input_eventX`) and a property name (`fixtures.foo_device_2_mouse`).
It also allows dynamic creation of new fixtures during runtime.

File: tests/test.py (hunk 3 above)
Comment (jonasBoss) [reply tests/test.py:403]: yes

Comment (sezanzeb) [reply tests/test.py:407]: isn't it redundant then?

Comment (jonasBoss) [reply tests/test.py:407]: yes it is. I created the property for convenience to improve the readability. The attribute name `dev_input_eventXX` allows us to access the attribute with `fixtures["/dev/input/eventXX"]` from the `__getitem__` method, which is necessary for backwards comparability with some tests and for dynamic fixtures.

Comment (sezanzeb) [reply tests/test.py:407]: ok

**Pull request merged**